README.md
# Harrison AI Dates Tech Task
> This module computes the number of days between two dates.

## Table of Contents
* [General Info](#general-information)
* [Technologies Used](#technologies-used)
* [Features](#features)
* [Setup](#setup)
* [Usage](#usage)
* [Room for Improvement](#room-for-improvement)
* [Acknowledgements](#acknowledgements)
* [Contact](#contact)


## General Information
The aim of the project is to compute the difference between two dates from scratch. I.e. without importing or otherwise building on existing packages for date processing.
Timezones were not considered and the dates were expected to be of format `YYYY-MM-DD` and based of the [Gregorian calendar](https://en.wikipedia.org/wiki/Gregorian_calendar).


## Technologies Used
- [Poetry](https://python-poetry.org/) - Python package and dependency management
- [Docker](https://www.docker.com/) - Containerisation tool
- [VSCode Remote Containers](https://code.visualstudio.com/docs/remote/containers) - Containerised development
- [GitHub Actions](https://github.com/features/actions) - CI/CD tool


## Features
- Python module to calculate date difference (`dates/date_diff.py`) with tests (`dates/tests/`)
- Dockerfile for reproducibility across dev and future 'prod' environments
- CI/CD pipeline (`.github/workflows/`) for automated testing & PyPI packaging


## Setup
1. System Dependencies
    - [Python^3.8][https://www.python.org/downloads/]
    - [Poetry](https://python-poetry.org/)
2. Install dependencies
    - `poetry install`
3. Activate poetry virtual environemnt
    - `poetry shell`


## Usage

### CLI Usage

```bash
python main.py --date1 "2012-01-10" --date2 "2012-01-11"
0
```

```bash
python main.py --date1 "2021-12-01" --date2 "2017-12-14"
1447
```

```bash
python main.py --date1 "20000-01-01" --date2 "2017-12-14"
ValueError: Year value must be of format YYYY for input 20000-01-01
```

```bash
python main.py --date1 "2022-02-29" --date2 "2017-12-14"
ValueError: YYYY-02-29 is only valid on a leap year for input 2022-02-29
```

### Docker Usage
```bash
docker build -t hai-dates .
docker run hai-dates --date1 "2012-01-10" --date2 "2012-01-11"
0
```

### Developer Usage
- Future features can be implemented by developing on a new branch (e.g. `dev`) and using the [Commitizen](https://commitizen-tools.github.io/commitizen/) tool to standardise commit messages.
- Tests and coverage can be run with `pytest --cov=dates`
- The `pre-commit-config.yaml` will run some basic code quality checks before commiting to GitHub, including
    - `black` - code formatting according to PEP8
    - `mypy` - static type checking
    - `pylint` - check code quality (e.g. programming errors or refactoring suggestions according to PEP8)
- A push to the `main` branch will run the GitHub Actions CI/CD pipeline where it will run the following jobs:
1. Quality - linting & run rests
2. Release - peforms [semantic versioning](https://python-semantic-release.readthedocs.io/en/latest/) on the repository by updating the version and applying it both the `pyproject.toml` file and a new Git tag version.
3. Publish - publishes a new package version to TestPyPI
- A push to main that doesn't contain `fix`, `feat` or `BREAKING CHANGE` in the commit message will only run the `Quality` job in the CI/CD pipeline. If these key words are present it will create a new release version (major=`BREAKING CHANGE`, minor=`feat`, patch=`fix`) and publish it to PyPI.

### Test PyPI
```bash
python3.8 -m pip install --index-url https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple/ dates

python3.8 -c 'from dates.date_diff import date_diff; num_days=date_diff(date1="2021-12-01", date2="2017-12-14"); print(num_days)'
```


## Room for Improvement
- Allow the module to accept different date formats (e.g. 'DD-MM-YYYY')
- Add code test coverage requirement to CI/CD pipeline
- Add docker container build & push step in CI/CD pipeline
- Developer documentation ( e.g. [Sphynx](https://www.sphinx-doc.org/en/master/))

## Acknowledgements
- Thanks Danny for setting this up! And Andy, Leo & Hong for reviewing :)


## Contact
Created by [@elmidelange](https://github.com/elmidelange) - feel free to contact me!


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.0.0 (2022-03-31)
### Feature
* Testing BREAKING CHANGE trigger ([`886d3ab`](https://github.com/elmidelange/hai-dates-tech-task/commit/886d3ab5b7e23f0e4318d0afadaa3557cb2016db))

### Breaking
*  ([`886d3ab`](https://github.com/elmidelange/hai-dates-tech-task/commit/886d3ab5b7e23f0e4318d0afadaa3557cb2016db))

## v0.2.6 (2022-03-31)
### Fix
* Testing fix release ([`cf98306`](https://github.com/elmidelange/hai-dates-tech-task/commit/cf98306e50323c148940ebacf32fa2705cbf650e))

## v0.2.5 (2022-03-31)
### Fix
* Testing CD trigger ([`32e7fb8`](https://github.com/elmidelange/hai-dates-tech-task/commit/32e7fb816c0d9b82a4b5f152f5862bb65bb06516))

## v0.2.4 (2022-03-31)
### Fix
* Testing CD workflow for publishing on new tag ([`da6cc9c`](https://github.com/elmidelange/hai-dates-tech-task/commit/da6cc9cfe599f8d7f1e4772b89150d821fe4c874))

## v0.2.3 (2022-03-28)
### Fix
* Adding loguru to prod requirements ([`7fe49ed`](https://github.com/elmidelange/hai-dates-tech-task/commit/7fe49edd4906ad9df8d6c432893454391fbc807f))

## v0.2.2 (2022-03-27)
### Fix
* Removing build.sh and renaming utils file ([`136e028`](https://github.com/elmidelange/hai-dates-tech-task/commit/136e02892d3ed2baad6f6d13c1f328af5bd87fa3))

## v0.2.1 (2022-03-27)
### Fix
* Moved loguru to dev requirements ([`c725013`](https://github.com/elmidelange/hai-dates-tech-task/commit/c725013ffb9c48839fd346b27d372bbe374ccabd))

## v0.2.0 (2022-03-27)
### Feature
* Documentation ([`013ccc3`](https://github.com/elmidelange/hai-dates-tech-task/commit/013ccc3ed3e303fd068dd2b6576042966601c9b9))

## v0.1.4 (2022-03-27)
### Fix
* Force pulling latest semantic release from main for Publish CI pipeline ([`ac7f2fc`](https://github.com/elmidelange/hai-dates-tech-task/commit/ac7f2fc91eb858838c1a402723fefe9f5277f341))

## v0.1.3 (2022-03-27)
### Fix
* Force pulling latest semantic reelase from repo in Publish CI pipeline ([`250c98d`](https://github.com/elmidelange/hai-dates-tech-task/commit/250c98df9a7906a9fd4db1657d3721a29b5fc32e))

## v0.1.2 (2022-03-27)
### Fix
* Seperated Publish into a new job in th CI pipeline ([`a652f93`](https://github.com/elmidelange/hai-dates-tech-task/commit/a652f932269fd4693fb07486a645a0b1e7ec3748))

## v0.1.1 (2022-03-27)
### Fix
* Pulling code in CI pipeline before publishing ([`12b294d`](https://github.com/elmidelange/hai-dates-tech-task/commit/12b294dd33b2eaf916ef5cab25cb12b309181029))

## v0.1.0 (2022-03-27)
### Feature
* First release of the dates package to TestPyPi ([`bfb8ec9`](https://github.com/elmidelange/hai-dates-tech-task/commit/bfb8ec94925f7d33a1fb263f458633eb62d350fc))


tests/test_utils.py
import pytest
from dates.utils import (
    check_format,
    count_leap_years,
    get_day_month_year,
    get_total_days,
    is_leap_year,
)


@pytest.mark.parametrize(
    "test_year,expected",
    [
        (2000, True),
        (2004, True),
        (2001, False),
        (1900, False),
    ],
)
def test_is_leap_year(test_year, expected):
    assert is_leap_year(test_year) == expected


@pytest.mark.parametrize(
    "test_input",
    [
        ("01-01-2020"),
        ("20000-01-01"),
        ("2022-02-29"),
        ("02022-02-22"),
        ("2022-002-22"),
        ("2022-02-022"),
        (20220222),
        ("2022-02-22T22:22:22"),
        ("2022-02-22-22:22:22"),
    ],
)
def test_check_format(test_input):
    with pytest.raises(ValueError):
        check_format(test_input)


@pytest.mark.parametrize(
    "test_input,expected",
    [
        ("2012-01-10", (2012, 1, 10)),
        ("9999-12-31", (9999, 12, 31)),
        ("0001-01-01", (1, 1, 1)),
    ],
)
def test_get_day_month_year(test_input, expected):
    assert get_day_month_year(test_input) == expected


@pytest.mark.parametrize(
    "test_year,test_month,test_day,expected",
    [
        (2000, 3, 1, 485),  # 97 / 400 years are leap years in the Gregorian calendar
        (2000, 2, 1, 484),
    ],
)
def test_count_leap_years(test_year, test_month, test_day, expected):
    assert count_leap_years(test_year, test_month) == expected


@pytest.mark.parametrize(
    "test_year,test_month,test_day,expected",
    [
        (2022, 1, 1, 738521),
        (0, 1, 1, 1),
        (3999, 12, 31, 1460969),
    ],
)
def test_get_total_days(test_year, test_month, test_day, expected):
    assert get_total_days(test_year, test_month, test_day) == expected


tests/test_date_diff.py
from dates.date_diff import date_diff


def test_no_days_diff():
    """Test for no days difference"""
    assert date_diff(date1="2012-01-10", date2="2012-01-11") == 0


def test_same_month():
    """Test for the same month"""
    assert date_diff(date1="2012-01-01", date2="2012-01-10") == 8


def test_different_month():
    """Test for different months"""
    assert date_diff(date1="1801-06-13", date2="1801-11-11") == 150


def test_different_year():
    """Test for different years"""
    assert date_diff(date1="2017-12-14", date2="2021-12-01") == 1447


def test_reverse_dates():
    """Test for reversed dates"""
    assert date_diff(date1="2021-12-01", date2="2017-12-14") == 1447


tests/test_date_utils.py
import pytest
from dates.dates_utils import (
    check_format,
    count_leap_years,
    get_day_month_year,
    get_total_days,
    is_leap_year,
)


@pytest.mark.parametrize(
    "test_year,expected",
    [
        (2000, True),
        (2004, True),
        (2001, False),
        (1900, False),
    ],
)
def test_is_leap_year(test_year, expected):
    assert is_leap_year(test_year) == expected


@pytest.mark.parametrize(
    "test_input",
    [
        ("01-01-2020"),
        ("20000-01-01"),
        ("2022-02-29"),
        ("02022-02-22"),
        ("2022-002-22"),
        ("2022-02-022"),
        (20220222),
        ("2022-02-22T22:22:22"),
        ("2022-02-22-22:22:22"),
    ],
)
def test_check_format(test_input):
    with pytest.raises(ValueError):
        check_format(test_input)


@pytest.mark.parametrize(
    "test_input,expected",
    [
        ("2012-01-10", (2012, 1, 10)),
        ("9999-12-31", (9999, 12, 31)),
        ("0001-01-01", (1, 1, 1)),
    ],
)
def test_get_day_month_year(test_input, expected):
    assert get_day_month_year(test_input) == expected


@pytest.mark.parametrize(
    "test_year,test_month,test_day,expected",
    [
        (2000, 3, 1, 485),  # 97 / 400 years are leap years in the Gregorian calendar
        (2000, 2, 1, 484),
    ],
)
def test_count_leap_years(test_year, test_month, test_day, expected):
    assert count_leap_years(test_year, test_month) == expected


@pytest.mark.parametrize(
    "test_year,test_month,test_day,expected",
    [
        (2022, 1, 1, 738521),
        (0, 1, 1, 1),
        (3999, 12, 31, 1460969),
    ],
)
def test_get_total_days(test_year, test_month, test_day, expected):
    assert get_total_days(test_year, test_month, test_day) == expected


app/date.py
from app.date_utils import check_format, get_day_month_year, get_total_days


def date_diff(date1: str, date2: str) -> int:
    """Returns the days difference between two dates.

    Args:
        date1 (str): The first date with format YYYY-MM-DD.
        date2 (str): The second date with format YYYY-MM-DD.

    Returns:
        int: Number of days difference
    """

    # Check for valid input format
    check_format(date_str=date1)
    check_format(date_str=date2)

    # Extract the year, month and day values
    y1, m1, d1 = get_day_month_year(date1)
    y2, m2, d2 = get_day_month_year(date2)

    # Get the total days since 0000-00-00
    total1 = get_total_days(year=y1, month=m1, day=d1)
    total2 = get_total_days(year=y2, month=m2, day=d2)

    # Get the absolute difference between the two dates
    diff = total2 - total1
    abs_diff = int((diff**2) ** (0.5))

    # Exclude the input dates by subtracting a day
    return abs_diff - 1


dates/dates.py
"""Dates module
"""
from dates.dates_utils import check_format, get_day_month_year, get_total_days


def date_diff(date1: str, date2: str) -> int:
    """Returns the days difference between two dates.

    Args:
        date1 (str): The first date with format YYYY-MM-DD.
        date2 (str): The second date with format YYYY-MM-DD.

    Returns:
        int: Number of days difference
    """

    # Check for valid input format
    check_format(date_str=date1)
    check_format(date_str=date2)

    # Extract the year, month and day values
    y_1, m_1, d_1 = get_day_month_year(date1)
    y_2, m_2, d_2 = get_day_month_year(date2)

    # Get the total days since 0000-00-00
    total1 = get_total_days(year=y_1, month=m_1, day=d_1)
    total2 = get_total_days(year=y_2, month=m_2, day=d_2)

    # Get the absolute difference between the two dates
    diff = total2 - total1
    abs_diff = int((diff**2) ** (0.5))

    # Exclude the input dates by subtracting a day
    return abs_diff - 1


dates/date_diff.py
"""Dates module
"""
from loguru import logger

from dates.utils import check_format, get_day_month_year, get_total_days


def date_diff(date1: str, date2: str) -> int:
    """Returns the days difference between two dates.

    Args:
        date1 (str): The first date with format YYYY-MM-DD.
        date2 (str): The second date with format YYYY-MM-DD.

    Returns:
        int: Number of days difference
    """
    logger.info(f"Calculating days between {date1} and {date2}")

    # Check for valid input format
    check_format(date_str=date1)
    check_format(date_str=date2)

    # Extract the year, month and day values
    y_1, m_1, d_1 = get_day_month_year(date1)
    y_2, m_2, d_2 = get_day_month_year(date2)

    # Get the total days since 0000-00-00
    total1 = get_total_days(year=y_1, month=m_1, day=d_1)
    total2 = get_total_days(year=y_2, month=m_2, day=d_2)

    # Get the absolute difference between the two dates
    diff = total2 - total1
    abs_diff = int((diff**2) ** (0.5))

    # Exclude the input dates by subtracting a day
    return abs_diff - 1


dates/dates_utils.py
"""Dates utilities file
"""

from typing import Tuple

from loguru import logger

month_days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]


def check_format(date_str: str) -> None:
    """Checks the format of the date string is valid (YYYY-MM-DD)

    Args:
        date (str): Date string
    """

    if not isinstance(date_str, str):
        msg = f"Date must be a string for input {date_str}"
        logger.error(msg)
        raise ValueError(msg)

    date_parts = date_str.split("-")

    if len(date_parts) != 3:
        msg = f"Date must be in YYYY-MM-DD format for input {date_str}"
        logger.error(msg)
        raise ValueError(msg)

    if not all(x.isdigit() for x in date_parts):
        msg = f"Date must only contain numbers for input {date_str}"
        logger.error(msg)
        raise ValueError(msg)

    year, month, day = tuple(int(x) for x in date_parts)

    if not 0 <= year <= 9999 or not len(date_parts[0]) == 4:
        msg = f"Year value must be of format YYYY for input {date_str}"
        logger.error(msg)
        raise ValueError(msg)

    if not 1 <= month <= 12 or not len(date_parts[1]) == 2:
        msg = f"Month value must be of format MM and be between 1 and 12 for input {date_str}"
        logger.error(msg)
        raise ValueError(msg)

    if not 0 <= day <= 31 or not len(date_parts[2]) == 2:
        msg = f"Day value must be of format DD and be between 1 and 31 for input {date_str}"
        logger.error(msg)
        raise ValueError(msg)

    if (is_leap_year(year) is False) and (month == 2) and (day == 29):
        msg = f"YYYY-02-29 is only valid on a leap year for input {date_str}"
        logger.error(msg)
        raise ValueError(msg)


def is_leap_year(year: int) -> bool:
    """Checks if year is a leap year.

    ASSUMPTION: To be a leap year, the year number must be divisible by four,
                except for end-of-century years, which must be divisible by 400.

    Args:
        year (int): Year

    Returns:
        bool: Is leap year or not
    """
    # Check if century year and divisble by 400
    if (year % 400 == 0) and (year % 100 == 0):
        return True

    # Check non-century year and divisible by 4
    if (year % 4 == 0) and (year % 100 != 0):
        return True

    return False


def get_day_month_year(date_str: str) -> Tuple[int, ...]:
    """Returns the day, month and year for a given date string in YYYY-MM-DD format

    Args:
        date_str (str): Date string in YYYY-MM-DD format

    Returns:
        Tuple[int, int, int]: Year, month and day values
    """
    return tuple(int(x) for x in date_str.split("-"))


def get_total_days(year: int, month: int, day: int) -> int:
    """Returns the total number of days in a date since 0001-01-01

    Args:
        day (int): Day value
        month (int): Month Value
        year (int): Year value

    Returns:
        int: Total number of days
    """
    # initialize count using years and day
    num_days = year * 365 + day

    # Add days for months in given date
    for i in range(0, month - 1):
        num_days += month_days[i]

    # Add a day for every leap year where there are 366 days
    num_days += count_leap_years(year, month)

    return num_days


def count_leap_years(year: int, month: int) -> int:
    """Returns the number of extra days incurred from leap years for a given date

    Args:
        year (int): Year value
        month (int): Month value
        day (int): Day value

    Returns:
        int: Number of leap years
    """
    # Check if the current year needs to be considered
    total_years = year
    if month <= 2:
        total_years -= 1

    # A leap year is a multiple of 4 or 400 (not 100)
    return int(total_years / 4) - int(total_years / 100) + int(total_years / 400)
